pow/src/main/java/tech/pegasys/artemis/pow/Eth1Manager.java

```diff
@@ -13,113 +13,121 @@
+    SafeFuture<EthBlock.Block> minGenesisBlock =
+        head.thenCompose(this::findMinGenesisTimeBlockInHistory);
+
+    return minGenesisBlock
+        .thenCompose(
+            block -> depositProcessingController.fetchDepositsFromGenesisTo(block.getNumber()))
+        .thenRun(
+            () -> {
+              EthBlock.Block block = checkNotNull(minGenesisBlock.getNow(null));
+              postMinGenesisTimeBlock(eth1EventsChannel, block);
+              depositProcessingController.startSubscription(block.getNumber());
+            });
   }
 
-  public SafeFuture<Void> findAndPublishFirstValidBlock() {
+  public SafeFuture<EthBlock.Block> getHead() {
     return eth1Provider
         .getLatestEth1BlockFuture()
         .thenApply(EthBlock.Block::getNumber)
         .thenApply(number -> number.subtract(Constants.ETH1_FOLLOW_DISTANCE.bigIntegerValue()))
         .thenApply(UnsignedLong::valueOf)
-        .thenCompose(eth1Provider::getEth1BlockFuture)
-        .thenCompose(
-            block -> {
-              int comparison = compareBlockTimestampToMinGenesisTime(block);
-              if (comparison > 0) {
-                // If block timestamp is greater than min genesis time
-                // find first valid block in history
-                return findFirstValidBlockInHistory(block);
-              } else if (comparison < 0) {
-                // If block timestamp is less than min genesis time
-                // subscribe to new block events and wait for the first
-                // valid block
-                return waitForFirstValidBlock();
-              } else {
-                return SafeFuture.completedFuture(block);
-              }
-            })
-        .thenAccept(this::publishFirstValidBlock)
-        .exceptionallyCompose(
-            err -> {
-              if (latestBlockDisposable != null) {
-                latestBlockDisposable.dispose();
-              }
+        .thenCompose(eth1Provider::getEth1BlockFuture);
+  }
 
-              LOG.debug(
-                  "Eth1MinGenesisTimeBlockFinder failed to find first valid block. Retry in "
-                      + Constants.ETH1_MIN_GENESIS_TIME_BLOCK_RETRY_TIMEOUT
-                      + " seconds",
-                  err);
-
-              System.out.println("adding one delayed run");
-              return asyncRunner.runAfterDelay(
-                  this::findAndPublishFirstValidBlock,
-                  Constants.ETH1_MIN_GENESIS_TIME_BLOCK_RETRY_TIMEOUT,
-                  TimeUnit.SECONDS);
-            });
+  public SafeFuture<Boolean> isHeadBeforeMinGenesis(SafeFuture<EthBlock.Block> headBlock) {
+    return headBlock.thenApply(
+        block -> {
+          int comparison = compareBlockTimestampToMinGenesisTime(block);
+          // If block timestamp is greater than min genesis time,
+          // min genesis block must have been in history
+          return comparison > 0;
```

Comment:
If it's equal then min genesis time is also in the existing block chain list (because it is the head block).

Reply:
oh true. good catch!